snapshot-first-storage.md

```diff
@@ -0,0 +1,148 @@
+
+Objects in s3 are immutable snapshots, never to be modified once written (only deleted).
+
+Objects in s3 are files, each containing a set of pages for some branch/relation/segment as of a specific time (LSN). A snapshot could be complete (meaning it has a copy of every page), or it could be incremental (containing only the pages that were modified since the previous snapshot). It’s expected that most snapshots are incremental to keep storage costs low.
+
+It’s expected that the pageserver would upload new snapshot objects frequently, e.g. somewhere between 30 seconds and 15 minutes, depending on cost/performance balance.
+
+No-longer needed snapshots can be “squashed”-- meaning snapshot N and snapshot N+1 can be read by some cloud agent software, which writes out a new object containing the combined set of pages (keeping only the newest version of each page) and then deletes the original snapshots.
+
+A pageserver only needs to store the set of pages needed to satisfy operations in flight: if a snapshot is still being written, the pageserver needs to hold historical pages so that snapshot captures a consistent moment in time (similar to what is needed to satisfy a slow replica).
+
+WAL records can be discarded once a snapshot has been stored to s3. (Unless we want to keep them longer as part of a point-in-time recovery feature.)
+
+## Pageserver operation
+
+To start a pageserver from a stored snapshot, the pageserver downloads a set of snapshots sufficient to start handling requests. We assume this includes the latest copy of every page, though it might be possible to start handling requests early, and retrieve pages for the first time only when needed.
+
+To halt a pageserver, one final snapshot should be written containing all pending WAL updates; then the pageserver and safekeepers can shut down.
+
+It’s assumed there is some cloud management service that ensures only one pageserver is active and servicing writes to a given branch.
+
+The pageserver needs to be able to track whether a given page has been modified since the last snapshot, and should be able to produce the set of dirty pages efficiently to create a new snapshot.
+
+The pageserver need only store pages that are “reachable” from a particular LSN. For example, a page may be written four times, at LSN 100, 200, 300, and 400. If no snapshot is being created when LSN 200 is written, the page at LSN 100 can be discarded. If a snapshot is triggered when the pageserver is at LSN 299, the pageserver must preserve the page from LSN 200 until that snapshot is complete. As before, the page at LSN 300 can be discarded when the LSN 400 pages is written (regardless of whether the LSN 200 snapshot has completed.)
```

Comment:
Regarding:

> If no snapshot is being created when LSN 200 is written, the page at LSN 100 can be discarded.
> ...
> As before, the page at LSN 300 can be discarded when the LSN 400 pages is written (regardless of whether the LSN 200 snapshot has completed.)

Is it completely unrealistic for compute node (Postgres) to ask Page@LSN100 even if it just wrote WAL records for Page@LSN200?

I thought that due to MVCC it is legal that some long transaction with an older snapshot can read older version of some page. And this is not that rare when a lot of concurrent transactions occur. So in our case, if Pageserver just drops all images of PageN with LSN < 200 when materializing/writing Page@LSN200, then on the next GetPage@LSN100 we should go to S3 and materialize this older page since the prior snapshot. If so, is it fine from the performance (latency) perspective?

Reply:
On the second thought, I mixed a concept of MVCC with different versions of **record** and different **pages**. So if we write a new record version we may create a new page for doing that, but these are completely different pages from the Pageserver perspective and everything is fine. Sorry for inconvenience :)

Reply:
There is actually no MVCC at page level. Long living transaction will still read page with most recent LSN (actually with last written LSN, but it doesn't matter).   So master will never try to read old page version even in case of presence of long living transactions. It may be true only for lagged replicas.

Reply:
> On the second thought, I mixed a concept of MVCC with different versions of **record** and different **pages**.

Yeah, MVCC is not an issue here.

You have a point though. Currently, the compute node *will* in fact always use a slightly lagging-behind LSN in the requests. It keeps track of the LSN of the last page that was evicted from the buffer cache, and uses that in the requests, instead of the LSN of the last-inserted WAL record as you might expect. That's an optimization. We know that the page being requested cannot have changed after the last-evicted LSN. If it was, it would either still be in the buffer cache and we would not be requesting it from the page server, or we would have updated the last-evicted LSN. That optimization allows the page server to respond immediately to the GetPage@LSN request even if it hasn't fully caught up to the latest generated LSN, reducing latency.

But that could be a problem if the page server had already garbage collected some page versions after the last-evicted LSN. Perhaps the GetPage@LSN request should include two LSNs, min and max. The semantics would be that the compute node knows that the page hasn't changed between the min and max LSN, and the page server is therefore allowed to return any page version in that range (there should be only one, we could make it an error if there are multiple). The compute node would use the last-evicted LSN as the min, and last-generated LSN as the max.

Alternatively, the GetPage@LSN request could carry a flag to indicate that a later LSN is OK as well. So the page server could respond with a page at the given LSN or later (and there should again be only one such version in practice).

Reply:
That is interesting, thanks for the explanation!